charmtools/build/tactics.py

```diff
@@ -750,35 +751,45 @@ def combine(self, existing):
         self.previous = existing.previous + [existing]
         return self
 
-    def _add(self, pip, wheelhouse, *reqs):
+    def _add(self, wheelhouse, *reqs):
         with utils.tempdir(chdir=False) as temp_dir:
             # put in a temp dir first to ensure we track all of the files
-            utils.Process((pip, 'install',
-                           '--no-binary', ':all:',
-                           '-d', temp_dir) +
-                          reqs).exit_on_error()()
+            self._pip('download', '--no-binary', ':all:', '-d', temp_dir,
+                      *reqs)
             for wheel in temp_dir.files():
                 dest = wheelhouse / wheel.basename()
                 dest.remove_p()
                 wheel.move(wheelhouse)
                 self.tracked.append(dest)
 
+    def _pip(self, *args):
+        assert self._venv is not None
+        # have to use bash to call pip to activate the venv properly first
+        utils.Process(('bash', '-c', ' '.join(
+            ('.', self._venv / 'bin' / 'activate', ';', 'pip3') + args
+        ))).exit_on_error()()
+
     def __call__(self, venv=None):
-        create_venv = venv is None
-        venv = venv or path(tempfile.mkdtemp())
-        pip = venv / 'bin' / 'pip3'
+        create_venv = self._venv is None
```

Comment:
This method does not use the "venv" parameter in this version of the code. I would recommend removing or keeping the logic as it was previously.